cmd/kubeadm/app/discovery/discovery.go

```diff
@@ -17,15 +17,72 @@ limitations under the License.
 package discovery
 
 import (
+	"crypto/tls"
 	"fmt"
+	"io/ioutil"
+	"net/http"
 
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubenode "k8s.io/kubernetes/cmd/kubeadm/app/node"
+
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
 	clientcmdapi "k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api"
 )
 
-func For(c kubeadmapi.Discovery) (*clientcmdapi.Config, error) {
+// For identifies and executes the desired discovery mechanism.
+func For(d kubeadmapi.Discovery) (*clientcmdapi.Config, error) {
 	switch {
+	case d.File != nil:
+		return runFileDiscovery(d.File)
+	case d.HTTPS != nil:
+		return runHTTPSDiscovery(d.HTTPS)
+	case d.Token != nil:
+		return runTokenDiscovery(d.Token)
 	default:
-		return nil, fmt.Errorf("unimplemented")
+		return nil, fmt.Errorf("Couldn't find a valid discovery configuration. Please provide one.")
+	}
+}
+
+// runFileDiscovery executes file-based discovery.
+func runFileDiscovery(fd *kubeadmapi.FileDiscovery) (*clientcmdapi.Config, error) {
+	return clientcmd.LoadFromFile(fd.Path)
+}
+
+// runHTTPSDiscovery executes HTTPS-based discovery.
+func runHTTPSDiscovery(hd *kubeadmapi.HTTPSDiscovery) (*clientcmdapi.Config, error) {
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```

Comment:
This needs to be secure. The idea is that we are pulling from an already trusted source (e.g. https://storage.googleapis.com). The doc on this method is here https://github.com/kubernetes/community/blob/master/contributors/design-proposals/bootstrap-discovery.md#method-https-endpoint